Five fixes for the perf graph live prints, and two bonus outputs improvements

GiudGiud opened the pull request:
perf graph needed some TLC.

This isnt the end of it, I dont think I am fixing some of our app prints interaction with the perf graph

GiudGiud committed:
Fix part of live prints for multiapps
- dont forget the size of the prefix before filling the line
refs #19114

GiudGiud committed:
Even if something else did not print, we need to skip a line before printing "Finished" on a section
closes #23744

GiudGiud committed:
Print total memory used, not differential tallies of memory, closes #23747

GiudGiud committed:
Avoid printing the perf graph message on every UO execution group, closes #23739

GiudGiud committed:
Add execution flag to perf graph print to try to make section more helpful

GiudGiud committed:
Differentiate between general output and checkpoint output in prints

GiudGiud committed:
Still print skipped sections otherwise we get unexplained dot dot dots, refs #23746

GiudGiud committed:
Better postprocessor table output, refs #22856
- make default environment based default to AUTO mode if fails
- if AUTO fails, default to 170, the size of Guillaume's terminal

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1407570/) on 8a17baa wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23752/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23752/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1f188dd3d7d67e657d0aa31546a71f0fd21d32b2`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1407570/) on 8a17baa wanted to post the following:

The following commit(s) contain fixups:

- e48c79e099: fixup! Print total memory used, not differential tallies of memory, closes #23747
- 227b3d8402: fixup! Avoid printing the perf graph message on every UO execution group, closes #23739

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1407570/) on 8a17baa wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23752/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23752/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1f188dd3d7d67e657d0aa31546a71f0fd21d32b2`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1407570/) on 8a17baa wanted to post the following:

The following commit(s) contain fixups:

- e48c79e099: fixup! Print total memory used, not differential tallies of memory, closes #23747
- 227b3d8402: fixup! Avoid printing the perf graph message on every UO execution group, closes #23739

GiudGiud commented:
Final fix isnt working yet. I might drop it for the sake of my time

GiudGiud commented:
Let's get those in.
The fix for tables is actually a big deal for SAM output

GiudGiud commented:
Before
```


'args' has been deprecated and will be removed on 02/07/2024. Please use 'coupled_variables' instead.

Setting Up.....................
  Setting Up Undisplaced Mesh
    Preparing Mesh
      Updating Mesh
        Finished Building Node List                                                      [  0.85 s] [  172 MB]
        Finished Building Boundary Elements List                                         [  0.70 s] [  196 MB]
      Still Updating Mesh..                                                              [ 17.70 s] [  928 MB]
    Finished Preparing Mesh                                                              [ 21.10 s] [ 1070 MB]
  Finished Setting Up Undisplaced Mesh                                                   [ 21.10 s] [ 1070 MB]
Still Setting UpKeeping file './tmp_jit_nzq3Mf.cc' in debug mode.
Keeping file './tmp_jit_ThmHbz.cc' in debug mode.
Keeping file './tmp_jit_I41QlU.cc' in debug mode.
Keeping file './tmp_jit_Q4QUgz.cc' in debug mode.
Keeping file './tmp_jit_0wjQvx.cc' in debug mode.
Keeping file './tmp_jit_TApgLQ.cc' in debug mode.
Keeping file './tmp_jit_xpH0XQ.cc' in debug mode.
Keeping file './tmp_jit_5m6qmd.cc' in debug mode.

  Initializing
    Updating Because Mesh Changed
      Updating Mesh..                                                                    [ 20.18 s] [ -336 MB]
    Finished Updating Because Mesh Changed                                               [ 22.16 s] [ -279 MB]
.................................                       [172.73 s] [-2606 MB]
  Finished Initializing                                                                  [195.41 s] [-2924 MB]
Finished Setting Up                                                                      [339.42 s] [ 3178 MB]

Framework Information:
MOOSE Version:           git commit 5f95f8e7bf on 2023-03-13
LibMesh Version:         
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Wed Mar 15 12:34:19 2023
Executable Timestamp:    Mon Mar 13 21:03:41 2023

Parallelism:
  Num Processors:          8
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
  Num Subdomains:          49
  Num Partitions:          8
  Partitioner:             metis

  TimeStepper:             IterationAdaptiveDT
  Solver Mode:             Preconditioned JFNK
  PETSc Preconditioner:    hypre boomeramg strong_threshold: 0.7 (auto)



    Finished Setting Up Outputs                                                          [  4.08 s] [  251 MB]
    Projecting Initial Solutions.........                                                [ 50.56 s] [  432 MB]

      Computing Initial Material Values..                                                [ 21.86 s] [ -153 MB]
    Finished Setting Up Materials                                                        [ 21.86 s] [ -153 MB]
    Computing User Objects..                                                             [ 21.03 s] [  -14 MB]
  Still Performing Initial Setup..
    Computing User Objects...

                                                            [ 22.19 s] [    0 MB]
  Finished Performing Initial Setup                                                      [141.27 s] [  422 MB]


Time Step 0, time = -10000

Postprocessor Values:
jaja
..............
    Outputting Step...............................                                       [165.64 s] [ -769 MB]
  Finished Outputting                                                                    [244.40 s] [  297 MB]
  Finished Advancing State                                                               [  0.32 s] [  308 MB]


Time Step 1, time = -9999, dt = 1

```

After
```
Setting Up.....................
  Setting Up Undisplaced Mesh
    Preparing Mesh
      Updating Mesh
        Finished Building Node List                                                      [  0.82 s] [ 7679 MB]
        Finished Building Boundary Elements List                                         [  0.67 s] [ 7847 MB]
      Still Updating Mesh.                                                               [ 15.68 s] [ 8354 MB]
    Finished Preparing Mesh                                                              [ 19.32 s] [ 8354 MB]
  Finished Setting Up Undisplaced Mesh                                                   [ 19.32 s] [ 8354 MB]
Still Setting Up
  Initializing
    Updating Because Mesh Changed
      Updating Mesh
        Finished Building Node List                                                      [  0.96 s] [ 8766 MB]
      Still Updating Mesh.                                                               [ 14.30 s] [ 6652 MB]
    Finished Updating Because Mesh Changed                                               [ 16.30 s] [ 6750 MB]
    Currently Initializing Equation Systems.............................                 [154.13 s] [ 4773 MB]
  Finished Initializing                                                                  [170.54 s] [ 4689 MB]
Finished Setting Up                                                                      [311.68 s] [ 4665 MB]

Framework Information:
MOOSE Version:           git commit 3c52016436 on 2023-03-15
LibMesh Version:         
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Wed Mar 15 22:18:26 2023
Executable Timestamp:    Wed Mar 15 22:06:28 2023

Parallelism:
  Num Processors:          8
  Num Threads:             1

  Approximation Orders:    "FIRST" "FIRST" "FIRST" "FIRST" "FIRST" "FIRST" "FIRST" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             IterationAdaptiveDT
  TimeIntegrator:          ImplicitEuler
  Solver Mode:             Preconditioned JFNK
  PETSc Preconditioner:    hypre boomeramg strong_threshold: 0.7 (auto)



Currently Executing
  Performing Initial Setup
    Projecting Initial Solutions........                                                 [ 49.62 s] [ 4625 MB]
    Currently Setting Up Materials
      Computing Initial Material Values..                                                [ 21.34 s] [ 4622 MB]
    Finished Setting Up Materials                                                        [ 21.34 s] [ 4622 MB]
    Computing User Objects..                                                             [ 20.60 s] [ 4655 MB]
  Still Performing Initial Setup..
    Computing User Objects...


    Finished Computing User Objects                                                      [ 21.24 s] [ 4653 MB]
  Finished Performing Initial Setup                                                      [137.25 s] [ 4653 MB]


Time Step 0, time = -10000



  Currently Outputting on INITIAL...........                                             [ 64.65 s] [ 5618 MB]
  Finished Advancing State                                                               [  0.18 s] [ 5986 MB]
```

GiudGiud commented:
Diff looks better on a real console

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1407735/) on 04299cc wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23752/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1407771/) on 04299cc wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>1f188d</th>
      <th colspan=4>#23752 04299c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.83%</td>
      <td>84.83%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>84017</td>
      <td>84026</td>
      <td>+9</td>
      <td>29</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15026</td>
      <td>15024</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23752/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23752/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23752/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
any ideas on that last one (the one not part of this PR) ?

```
      Computing Residual..

*** Warning ***
jaja le warning
....                                                           [ 38.65 s] [ 5536 MB]
```
It s tough catching that a warning was output, as it goes straight to Moose::out instead of the Console ?

GiudGiud commented:
I think the last think is to get MooseWarning to use the Console. I ll leave that out for now if you dont mind

lindsayad commented on framework/src/utils/FormattedTable.C:
```diff
@@ -609,7 +609,8 @@ FormattedTable::getTermWidth(bool use_environment) const
   }
-  else
+  // Default to AUTO if no environment variable was set
+  if (!use_environment || (w.ws_col == std::numeric_limits<unsigned short>::max()))
```
If `use_environment` is `false`, then `w.ws_col` is guaranteed to be `std::numeric_limits<unsigned short>::max()`, right? Anyway I think we can simplify this
```suggestion
  if (w.ws_col == std::numeric_limits<unsigned short>::max())
```

GiudGiud replied:
that works

lindsayad commented on framework/src/utils/PerfGraphLivePrint.C:
```diff
@@ -135,10 +143,14 @@ PerfGraphLivePrint::printStats(PerfGraph::SectionIncrement & section_increment_s
 
+  // Add multiapps prefix size
+  if (!_app.isUltimateMaster())
+    num_horizontal_chars += _app.name().size() + 2;
```
Is this hard-coded two going to work for nested mulitiapps?

GiudGiud replied:
yes it does! even added a test for it

lindsayad reviewed (commented):
Have you been seeing random failures?

lindsayad commented on framework/src/utils/FormattedTable.C:
```diff
@@ -619,11 +620,13 @@ FormattedTable::getTermWidth(bool use_environment) const
 
+  // Something bad happened, make sure we have a sane value
+  if (w.ws_col == std::numeric_limits<unsigned short>::max())
+    w.ws_col = 170;
```
😠

GiudGiud replied:
it's a good value!

lindsayad replied:
Maybe we should go off some standard presets. 80 is of course the standard. But GNOME has some 132 presets as well (https://help.gnome.org/users/gnome-terminal/stable/app-terminal-sizes.html.en)

GiudGiud replied:
we have bigger screens than that these days though

lindsayad replied:
On the subject of commit messages, could you please keep the length of your initial commit line under 80 characters? When people look at short logs your initial commit lines always wrap across the screen. (In this example the length of your name doesn't do the commit message any services, but you get the idea). Basically if GitHub has to wrap the initial commit line, then it's too long

<img width="832" alt="Screen Shot 2023-03-16 at 10 31 48 AM" src="https://user-images.githubusercontent.com/10248304/225704429-52111aa1-9dd1-4b19-8d97-0810b06e6917.PNG">

GiudGiud replied:
I ll try to remember. Didnt know about it

GiudGiud committed:
Add test for perf graph live print formatting and multiapps interaction, refs #19114 #23746

GiudGiud committed:
Add more timed section functionalities to the SlowProblem test object

GiudGiud committed:
Perf graph rightfully does not add a line break after some of its messages
to be able to print dot dot dots
But this confuses the multiapp prefix print, so let s add some logic to make sure we always get the prefix printed
refs #19114

GiudGiud commented:
No random failures and the new tests are quite failure resistant imo

GiudGiud committed:
Apply suggestions from code review
- use 132 for default console size
- simplify logic

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
should be ready @lindsayad

loganharbour commented:
You requested my review - are you not looking for it now?

loganharbour commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -4699,7 +4699,7 @@ FEProblemBase::execMultiApps(ExecFlagType type, bool auto_advance)
   if (multi_apps.size())
   {
-    TIME_SECTION("execMultiApps", 1, "Executing MultiApps", false);
+    TIME_SECTION("execMultiApps", 1, "Executing MultiApps on " + type.name(), false);
```
We simply cannot do this with the static registry. The name must be static.

GiudGiud replied:
fixed

loganharbour replied:
And for context in case someone comes across this later - because the registry is static, you only get one ID per instantiation in code. Which means everything that runs from that point will take the first statically registered name. Later names will do nothing

GiudGiud committed:
Address Logan's review:
- dont use timed section registry with dynamic names
- revert to old timed section names

moosebuild commented:
Job [Docker](https://civet.inl.gov/job/1411460/) on a80afe0 : invalidated by @GiudGiud

fluke

moosebuild commented:
Job [Docker](https://civet.inl.gov/job/1411460/) on a80afe0 : invalidated by @lindsayad

lindsayad commented:
If this fails a third time, then I'm wondering whether you may have actually slowed things down?

lindsayad commented on framework/src/utils/FormattedTable.C:
```diff
@@ -619,11 +620,13 @@ FormattedTable::getTermWidth(bool use_environment) const
 
+  // Something bad happened, make sure we have a sane value
+  if (w.ws_col == std::numeric_limits<unsigned short>::max())
+    w.ws_col = 132;
```
Put a comment about why this number was selected

lindsayad commented on test/tests/utils/perf_graph_live_print/tests:
```diff
@@ -1,11 +1,66 @@
+
+    issues = '#15444'
+    design = 'utils/PerfGraphLivePrint.md'
+    requirement = 'The system shall allow for nested timing sections of code and having automated print-outs when they take too long'
```
```suggestion
    requirement = 'The system shall allow for nested code timing sections and having automated print-outs of the code section names when they are taking a long time.'
```

lindsayad commented on test/tests/utils/perf_graph_live_print/tests:
```diff
@@ -1,11 +1,66 @@
+
+    issues = '#23746'
+    design = 'utils/PerfGraphLivePrint.md'
+    requirement = 'The system shall be able to print the automatic print-out of a timed section when an external print happened'
```
This requirement isn't very clear. We shall be able to print a print-out?

GiudGiud replied:
reworded

lindsayad commented on test/tests/utils/perf_graph_live_print/tests:
```diff
@@ -1,11 +1,66 @@
+
+    issues = '#19114'
+    design = 'utils/PerfGraphLivePrint.md'
+    requirement = 'The system shall allow for automatic print-outs of timing sections within multiapps'
```
```suggestion
    requirement = 'The system shall allow for automatic print-outs of timing sections within multiapps.'
```

lindsayad commented on test/tests/utils/perf_graph_live_print/tests:
```diff
@@ -1,11 +1,66 @@
+
+    issues = '#19114'
+    design = 'utils/PerfGraphLivePrint.md'
+    requirement = 'The system shall allow for automatic print-outs of nested timing sections within multiapps'
```
```suggestion
    requirement = 'The system shall allow for automatic print-outs of nested timing sections within multiapps.'
```

lindsayad commented on test/tests/utils/perf_graph_live_print/tests:
```diff
@@ -1,11 +1,66 @@
+
+    issues = '#19114'
+    design = 'utils/PerfGraphLivePrint.md'
+    requirement = 'The system shall allow for automatic print-outs of timing sections within nested multiapps'
```
```suggestion
    requirement = 'The system shall allow for automatic print-outs of timing sections within nested multiapps.'
```

GiudGiud commented on framework/src/utils/FormattedTable.C:
```diff
@@ -619,11 +620,13 @@ FormattedTable::getTermWidth(bool use_environment) const
     }
   }
 
+  // Something bad happened, make sure we have a sane value
```
```suggestion
  // Something bad happened, make sure we have a sane value
  // 132 seems good for medium sized screens, and is available as a GNOME preset
```

GiudGiud committed:
Apply suggestions from code review

dots at the end of test requirements
reword a test requirement
add comment on default terminal width

Also make test a tad longer to trigger pref live print more reliably

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
seems I need to make the multiapps test slower
https://civet.inl.gov/job/1414514/

GiudGiud commented:
Added a second to the test, adding up to 3s over nested multiapps

GiudGiud commented:
Thanks for the reviews!

@lingzou This should make the output of many postprocessors (such as in SAM) by default (eg without the environment variables) look a lot better on the console. Please hit me up if it's not as expected

loganharbour commented:
I fear this test might not be deterministic: https://civet.inl.gov/job/1415567/

GiudGiud commented:
Seems like a formatting problem again
```
utils/perf_graph_live_print.multiapps: subapp0: Testing Slowness. 0 Nonlinear |R| = 3.082207e+00
```

GiudGiud commented:
It's one of the ones I did not address, printouts that happen during a dotdotdot phase. 

It's the same problem for Griffin printouts that end up starting with a dot

I dont have a solution at this point, except maybe deciding that the console should skip a line if it has last printed a dot.
